be/src/connector/lake_connector.cpp

```diff
@@ -0,0 +1,598 @@
+
+class LakeDataSourceProvider final : public DataSourceProvider {
+public:
+    friend class LakeDataSource;
+    LakeDataSourceProvider(vectorized::ConnectorScanNode* scan_node, const TPlanNode& plan_node);
+    ~LakeDataSourceProvider() override = default;
+
+    DataSourcePtr create_data_source(const TScanRange& scan_range) override;
+
+protected:
+    vectorized::ConnectorScanNode* _scan_node;
+    const TOlapScanNode _t_olap_scan_node;
+};
+
+// ================================
+
+LakeDataSource::LakeDataSource(const LakeDataSourceProvider* provider, const TScanRange& scan_range)
+        : _provider(provider), _scan_range(scan_range.internal_scan_range) {}
+
+LakeDataSource::~LakeDataSource() {
+    _reader.reset();
+    _predicate_free_pool.clear();
+}
+
+Status LakeDataSource::open(RuntimeState* state) {
+    _runtime_state = state;
+    const TOlapScanNode& thrift_olap_scan_node = _provider->_t_olap_scan_node;
+    const TupleDescriptor* tuple_desc = state->desc_tbl().get_tuple_descriptor(thrift_olap_scan_node.tuple_id);
+    _slots = &tuple_desc->slots();
+
+    _runtime_profile->add_info_string("Table", tuple_desc->table_desc()->name());
+    _runtime_profile->add_info_string("IsLakeTable", "true");
+    if (thrift_olap_scan_node.__isset.rollup_name) {
+        _runtime_profile->add_info_string("Rollup", thrift_olap_scan_node.rollup_name);
+    }
+    if (thrift_olap_scan_node.__isset.sql_predicates) {
+        _runtime_profile->add_info_string("Predicates", thrift_olap_scan_node.sql_predicates);
+    }
+
+    init_counter(state);
+
+    // eval const conjuncts
+    RETURN_IF_ERROR(OlapScanConjunctsManager::eval_const_conjuncts(_conjunct_ctxs, &_status));
+
+    // Init _conjuncts_manager.
+    OlapScanConjunctsManager& cm = _conjuncts_manager;
+    cm.conjunct_ctxs_ptr = &_conjunct_ctxs;
+    cm.tuple_desc = tuple_desc;
+    cm.obj_pool = &_obj_pool;
+    cm.key_column_names = &thrift_olap_scan_node.key_column_name;
+    cm.runtime_filters = _runtime_filters;
+    cm.runtime_state = state;
+
+    const TQueryOptions& query_options = state->query_options();
+    int32_t max_scan_key_num;
+    if (query_options.__isset.max_scan_key_num && query_options.max_scan_key_num > 0) {
+        max_scan_key_num = query_options.max_scan_key_num;
+    } else {
+        max_scan_key_num = config::doris_max_scan_key_num;
+    }
+    bool enable_column_expr_predicate = false;
+    if (thrift_olap_scan_node.__isset.enable_column_expr_predicate) {
+        enable_column_expr_predicate = thrift_olap_scan_node.enable_column_expr_predicate;
+    }
+
+    // Parse conjuncts via _conjuncts_manager.
+    RETURN_IF_ERROR(cm.parse_conjuncts(true, max_scan_key_num, enable_column_expr_predicate));
+
+    RETURN_IF_ERROR(build_scan_range(_runtime_state));
+
+    RETURN_IF_ERROR(init_tablet_reader(_runtime_state));
+    return Status::OK();
+}
+
+void LakeDataSource::close(RuntimeState* state) {
+    update_counter();
+    _prj_iter->close();
```

Comment:
Please check these pointers before close.

If open failed, these pointers would be `nullptr` and cause a crash here.

Reply:
done